refactor: remove typed oracle and txe typed oracle

nventuro opened the pull request:
This PR removes `TypedOracle` and `TXETypedOracle`. In their place, it creates 5 interfaces for the different kinds of oracles:
 - utility
 - private
 - avm
 - txe
 - misc (a new kind, used for stuff that's available both in utility and also top level tests, such as random())

These interfaces are then implemented by the different classes that handle execution. The interfaces include runtime markers so that we can figure out whether a generic handler implements a specific interface or not, and throw during foreign call translation if the current handler doesn't support it (e.g. when calling a private oracle in the top level context in a test, or when calling a private oracle while simulating a utility function). We used to be able to always call these functions and then have their implementation error out via the default `TypedOracle` implementation, but this was a bit messy as there was no actual way to check for full interface implementation.

In doing this I found some minor issues with the oracles, which already make this be valuable, e.g. there were some abandoned oracles that were just never invoked but still declared in the TXETypedOracle object. I'll do some oracle cleanup later on, but not now in order to try to make all the changes happen in one go and reduce noise there.

The **real** reason why I made this change is because now `txe_oracle` can be replaced by `private_execution_oracle` and `utility_execution_oracle`, since they now implement the same interface. This will mean that Noir tests will be using the real PXE implementation and not a copy of it. I'll do this in a separate PR because there'll likely be things that need to be adjusted there.

The diff is a bit messy due to interface changes combined with files being renamed. I did not actually change any of the oracle methods.

benesjan commented on noir-projects/aztec-nr/aztec/src/test/helpers/authwit.nr:
```diff
@@ -12,18 +11,20 @@ use protocol_types::{
 where
     C: CallInterface<M>,
 {
-    // Note that in top-level context only chain_id and version are populated on the utility context so make sure to
-    // not use the other fields. This is a tech debt.
-    let utility_context = get_utility_context();
-    let chain_id = utility_context.chain_id();
-    let version = utility_context.version();
```
🥳

benesjan commented on noir-projects/aztec-nr/aztec/src/test/helpers/authwit.nr:
```diff
@@ -12,18 +11,20 @@ use protocol_types::{
-    let utility_context = get_utility_context();
-    let chain_id = utility_context.chain_id();
-    let version = utility_context.version();
+    // The creation of this utility context results in the TXE session needing to keep track of authwits over its
+    // lifetime, as we'd otherwise be unable to store more than one authwit due to them getting reset when restoring the
+    // top-level TXE context.
```
Why do you think this matters? Is it so memory intensive or smt.?

nventuro replied:
Why it matters that it is sad that we need to do it? Because it's not very clean, it breaks the separation of concerns from the session and the handlers a bit. Ultimately creating authwits on the top level context is something we'd ideally not do at all.

benesjan commented on yarn-project/pxe/src/bin/check_oracle_version.ts:
```diff
@@ -20,18 +16,35 @@ class OracleMock extends TypedOracle {}
+    .sort()
+    .join('');
+
   // Create a hashable representation of the oracle interface by concatenating its method names. Return values are
   // excluded from the hash calculation since they are typically arrays of fields and I didn't manage to reliably
   // stringify them.
```
```suggestion
```

This comment is now stale.

benesjan commented on yarn-project/pxe/src/bin/check_oracle_version.ts:
```diff
@@ -20,18 +16,35 @@ class OracleMock extends TypedOracle {}
+  // TODO: we're only checking the functions implemented by the Oracle object, which is really an ACVM translator, akin
+  // to TXE's RPC translator. This is correct in that it is ultimately the foreign interface that matters, but incorrect
+  // in that it does not take into consideration TXE's extended interface. An improvement would be to check
+  // RPCTranslator from TXE, though that still seems a bit fragile.
```
Would merge the comment on line 14 with this one and keep the link to the issue "TODO(#16581)".

benesjan commented on yarn-project/txe/src/oracle/txe_oracle.ts:
```diff
@@ -416,4 +406,56 @@ export class TXE extends TXETypedOracle {
     return txEffect;
   }
+
+  // TODO: this class will soon be replaced with the real UtilityExecutionOracle and PrivateExecutionOracle classes. The
+  // functions below are not currently used in Noir tests, and in most cases they're caught beforehand by the RPC
+  // translator - we just have a temporary empty implementation until we finalize the migration.
```
Thanks for this comment. I was always so confused by this class.

benesjan reviewed (approved):
Very nice

Thunkar commented:
Love the approach, hate the `IClass` naming convention for interfaces. I don't have a better suggestion though, but the java convention gives me the ick

nventuro committed:
refactor: remove typed oracle and txe typed oracle

This PR removes `TypedOracle` and `TXETypedOracle`. In their place, it creates 5 interfaces for the different kinds of oracles:
 - utility
 - private
 - avm
 - txe
 - misc (a new kind, used for stuff that's available both in utility and also top level tests, such as random())

These interfaces are then implemented by the different classes that handle execution. The interfaces include runtime markers so that we can figure out whether a generic handler implements a specific interface or not, and throw during foreign call translation if the current handler doesn't support it (e.g. when calling a private oracle in the top level context in a test, or when calling a private oracle while simulating a utility function). We used to be able to always call these functions and then have their implementation error out via the default `TypedOracle` implementation, but this was a bit messy as there was no actual way to check for full interface implementation.

In doing this I found some minor issues with the oracles, which already make this be valuable, e.g. there were some abandoned oracles that were just never invoked but still declared in the TXETypedOracle object. I'll do some oracle cleanup later on, but not now in order to try to make all the changes happen in one go and reduce noise there.

The **real** reason why I made this change is because now `txe_oracle` can be replaced by `private_execution_oracle` and `utility_execution_oracle`, since they now implement the same interface. This will mean that Noir tests will be using the real PXE implementation and not a copy of it. I'll do this in a separate PR because there'll likely be things that need to be adjusted there.

The diff is a bit messy due to interface changes combined with files being renamed. I did not actually change any of the oracle methods.

Co-authored-by: Nicolás Venturo <nicolas.venturo@gmail.com>
